Disable util::Result copying and assignment

ryanofsky opened the pull request:
This PR just contains the first two commits of #25665.

It disables copying of `util::Result` objects because unnecessary copies are inefficient and not possible after #25665, which makes `util::Result` object move-only.

It disables the assignment operator and replaces it with an `Update()` method, because #25665 adds more information to `util::Result` objects (warning and error messages and failure values) and having an assignment operator that overwrites data instead of merging it would make it easy to accidentally erase existing information while trying to assign new information.

ryanofsky committed:
refactor: Drop util::Result operator=

`util::Result` objects are aggregates that can hold multiple fields with
different information. Currently Result objects can only hold a success value
of an arbitrary type or a single bilingual_str error message. In followup PR
https://github.com/bitcoin/bitcoin/pull/25722, Result objects may be able to
hold both success and failure values of different types, plus error and warning
messages.

Having a Result::operator= assignment operator that completely erases all
existing Result information before assigning new information is potentially
dangerous in this case. For example, code that looks like it is assigning a
warning value could erase previously-assigned success or failure values.
Conversely, code that looks like it is just assigning a success or failure
value could erase previously assigned error and warning messages.

To prevent potential bugs like this, disable Result::operator= assignment
operator.

It is possible in the future we may want to re-enable operator= in limited
cases (such as when implicit conversions are not used) or add a Replace() or
Reset() method that mimicks default operator= behavior. Followup PR
https://github.com/bitcoin/bitcoin/pull/25722 also adds a Result::Update()
method providing another way to update an existing Result object.

Co-authored-by: stickies-v <stickies-v@protonmail.com>

ryanofsky committed:
refactor: Avoid copying util::Result values

Copying util::Result values is less efficient than moving them because they
allocate memory and contain strings. Also this is needed to avoid compile
errors in https://github.com/bitcoin/bitcoin/pull/25722 which adds a
std::unique_ptr member to util::Result which implicity disables copying.

stickies-v reviewed (commented):
Concept ACK, breaking it out in a smaller PR seems like a good idea, thanks.

I'm unsure about removing the assignment operator altogether. I think it's good to be careful about footguns, but I'm worried about making the interface overly unintuitive.

Would it make sense to only merge another `Result` when using the `Merge` (currently called `Update`) method, and use the assignment operator to overwrite the current result? I think when people want to merge results, they'll probably want to be explicit about that, and in other cases overwriting/clearing everything seems like expected behaviour?

ryanofsky reviewed (commented):
Updated 974b16b2580f4f09c8936fbead96fdbee637cb14 -> c45cb13b9e4f6eae50ab6dc42acf471921980591 ([`pr/saferesult.1`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.1) -> [`pr/saferesult.2`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.1..pr/saferesult.2)) with suggestions

re: https://github.com/bitcoin/bitcoin/pull/29906#pullrequestreview-2011883406

> I'm unsure about removing the assignment operator altogether. I think it's good to be careful about footguns, but I'm worried about making the interface overly unintuitive.

We could add back a safer, more limited form of operator= in the future if not having it causes friction. I rewrote commit message to be clearer and mention that. But I think having a named `Update()` method should make it easier to figure out what code is doing and avoid mistakes.

> Would it make sense to only merge another `Result` when using the `Merge` (currently called `Update`) method, and use the assignment operator to overwrite the current result? I think when people want to merge results, they'll probably want to be explicit about that, and in other cases overwriting/clearing everything seems like expected behaviour?

Hmm, I'm actually not sure what this is suggesting. If there is a need to overwrite/clear results, IMO, the best way to do that would be to add a `Reset()` method. I haven't seen an actual need to do that anywhere though. The intended use for Update is to be able to write code like:

```c++
util::Result<SuccessType, FailureType> result;
if (do_something()) {
  result.Update(success_value);
} else {
  result.Update({util::Error{_("error message"), failure_value});
}
```

ryanofsky reviewed (commented):
Updated c45cb13b9e4f6eae50ab6dc42acf471921980591 -> 66022315641934bda301d61f009dae9cb3b45da0 ([`pr/saferesult.2`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.2) -> [`pr/saferesult.3`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.2..pr/saferesult.3)) adding change to actually disable the copy constructor

sedited reviewed (approved):
ACK 66022315641934bda301d61f009dae9cb3b45da0

furszy reviewed (commented):
ACK 6602231564

stickies-v reviewed (commented):
Approach ACK, code LGTM 66022315641934bda301d61f009dae9cb3b45da0 with some non-blocking suggestions.

> We could add back a safer, more limited form of operator= in the future if not having it causes friction.

I made my previous comment because I thought there are some instances of somewhat awkward `Update` usage in this PR. The majority of `util::Result` usage is/should probably not require `Update()` at all, because we're just returning either a success value or an error string. But, in those simple cases, we should just use list initialization, imo, and not assignment initialization. So, I've left a couple of code comments that remove all the instances I find awkward.

> The intended use for Update is to be able to write code like: ...

Hmmm. In the code example you give, I'd find assignment much more intuitive, since the goal doesn't seem to be combining results. If `Update()` is aimed to be used exclusively (?) when we intend to combine results, doesn't that make for a more clear interface?

ryanofsky commented:
Thanks @stickies-v, I agree with your suggestions to avoid updating existing `Result` objects where possible and will apply them.

But I wonder if there are is anything else I can do to help make the interface more intuitive though. I'm not exactly clear on why it is not intuitive currently. A few places you mentioned "Update here is confusing" but I don't understand why it is confusing. Saying `result.Update(success_value)` is just meant to update the result object with a success value while keeping other result data intact. Saying `result.Update({util::Error{_("error message"), failure_value});` is meant to update the result object with an error message and failure value, again keeping other result data intact.

I think it's very similar to python dict.update function where if you start with an existing dictionary:

```python
d = {'a': 1, 'b': 2, 'c': 3}
```

and call the update method with new values:

```python
d.update({'a': 4, 'd': 5})
```

And the dictionary will be updated with the new values overwriting the old ones:

```python
{'a': 4, 'b': 2, 'c': 3, 'd': 5}
```

> In the code example you give, I'd find assignment much more intuitive, since the goal doesn't seem to be combining results. If Update() is aimed to be used exclusively (?) when we intend to combine results, doesn't that make for a more clear interface?

Update is just intended to update some data in a result object without necessarily erasing other data in the result object. Assigning is conceptually different and would imply that result being assigned to would be cleared and completely replaced by value being assigned. In this PR, the result object only has two mutually exclusive fields (success value and error string), so assigning and updating are the same thing. In #25665, result objects contain more fields so assigning and updating are different things.

ryanofsky commented on src/init.cpp:
```diff
@@ -993,7 +993,7 @@ bool AppInitParameterInteraction(const ArgsManager& args)
     auto result = init::SetLoggingCategories(args);
     if (!result) return InitError(util::ErrorString(result));
-    result = init::SetLoggingLevel(args);
+    result.Update(init::SetLoggingLevel(args));
```
re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1580022711

Thanks, applied patch

ryanofsky commented on src/test/mempool_tests.cpp:
```diff
@@ -260,7 +260,7 @@ BOOST_AUTO_TEST_CASE(MempoolIndexingTest)
     tx10.vout[0].nValue = 10 * COIN;
 
-    ancestors_calculated = pool.CalculateMemPoolAncestors(entry.Fee(200000LL).Time(NodeSeconds{4s}).FromTx(tx10), CTxMemPool::Limits::NoLimits());
+    ancestors_calculated.Update(pool.CalculateMemPoolAncestors(entry.Fee(200000LL).Time(NodeSeconds{4s}).FromTx(tx10), CTxMemPool::Limits::NoLimits()));
```
re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1580027359

Thanks, applied patch

ryanofsky commented on src/wallet/test/fuzz/notifications.cpp:
```diff
@@ -108,9 +108,9 @@ struct FuzzedWallet {
+            op_dest.Update(wallet->GetNewDestination(type, ""));
         } else {
-            op_dest = wallet->GetNewChangeDestination(type);
+            op_dest.Update(wallet->GetNewChangeDestination(type));
```
re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1580037125

Thanks, applied patch

ryanofsky commented on src/wallet/test/spend_tests.cpp:
```diff
@@ -102,7 +102,7 @@ BOOST_FIXTURE_TEST_CASE(wallet_duplicated_preset_inputs_test, TestChain100Setup)
     // Second case, don't use 'subtract_fee_from_outputs'.
     recipients[0].fSubtractFeeFromAmount = false;
-    res_tx = CreateTransaction(*wallet, recipients, /*change_pos=*/std::nullopt, coin_control);
+    res_tx.Update(CreateTransaction(*wallet, recipients, /*change_pos=*/std::nullopt, coin_control));
```
re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1580043133

Thanks, applied patch

ryanofsky reviewed (commented):
Updated 66022315641934bda301d61f009dae9cb3b45da0 -> 4fb08c28ba0f14782fdab4ac54387aae0aba82ed ([`pr/saferesult.3`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.3) -> [`pr/saferesult.4`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.3..pr/saferesult.4)) with suggested changes using new result objects to avoid updating existing ones where possible
Update 4fb08c28ba0f14782fdab4ac54387aae0aba82ed -> 88aa093356ff142efe60df989323eb1ed1aa9e18 ([`pr/saferesult.4`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.4) -> [`pr/saferesult.5`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.4..pr/saferesult.5)) with no changes overall, just moving changes out of first commit that belonged in second commit

stickies-v commented:
Sorry, I've been doing a poor job explaining myself. I am fully onboard with how `Update` works.  When I used the word "confusion", I was talking about _usage_, not interface*. Prior to this push, `Update()` was used in a couple of places where we know or expect the `Result` to be empty, and in those case I think we should just be explicit about that and use list initialization. When I see `result.Update()` being called, I have to look up which values are in `result` already and ensure that's correct, whereas with `auto result{...};` I know right away that we're starting from an empty result.

To stick with the python analogy: using `update` when we know or expect `a` to be an empty dictionary, is imo an anti-pattern and should be avoided.

```python
a = {}
// do some logic to calculate `value`
a.update({1: value})
```

Instead, it is much clearer to use assignment, so the reader doesn't have to figure out which values `a` held already.
```python
// do some logic to calculate `value`
a = {1: value}
```

In short: I think it would be helpful if we only use `result.Update()` when `result` (potentially) has had fields set already, and use list initialization (or if necessary assignment initialization) for when we know (and thus enforce that) `result` is empty. This is also the concern I had with the code example [at the end of this comment](https://github.com/bitcoin/bitcoin/pull/29906#pullrequestreview-2012191090).

I hope I did a better job explaining myself, but otherwise I'm happy to talk about it more offline to not pollute this PR more.

* well, except for making sure the method names align with how we intend people to use them, hence my earlier renaming suggestions.

ryanofsky commented:
Rebased 88aa093356ff142efe60df989323eb1ed1aa9e18 -> 0cbcb66257db65f0a68d82d17a2d1a83d8654b38 ([`pr/saferesult.5`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.5) -> [`pr/saferesult.6`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.5-rebase..pr/saferesult.6)) dropping Result::Update() method in this PR, and last 2 uses of it.

re: https://github.com/bitcoin/bitcoin/pull/29906#issuecomment-2078213252

Thanks for clarifying. I think we might disagree whether certain uses of an update function would be antipatterns, but I completely agree with all the code changes you suggested, and that in general if you have a choice between initializing values once or letting them change over time, you should prefer initializing them once so there's less changing state to reason about. I also looked at remaining two update calls in this PR, and thought they both were not good uses, so I removed them as well.

My [update example](https://github.com/bitcoin/bitcoin/pull/29906#pullrequestreview-2012191090]) was very simple because I wanted to show minimal code that demonstrated _how_ to use the update function, but the code was probably too simple to demonstrate _why_ you would want to use it. The use-case for update is when you have a function that is performing multiple operations and you want to accumulate warnings and errors from all the operations and return them. Examples would be the [`BlockManager::SaveBlockToDisk`](https://github.com/ryanofsky/bitcoin/blob/4d2c9de24916f8d69514ea7c7251136e2762fa5c/src/node/blockstorage.cpp#L1167-L1195) function from #29700 or the [`CompleteChainstateInitialization`](https://github.com/ryanofsky/bitcoin/blob/55d7de92bbbe035c1833c89f885af14e5b243932/src/node/chainstate.cpp#L38-L179) function from #25665. I think these are reasonable uses of the update function, but it's possible there could be better ways to write this code, and you might have another idea what it should look like.

stickies-v reviewed (approved):
ACK e1be443315be6ba6c80267e0e6be59deee77de50

> The use-case for update is when you have a function that is performing multiple operations and you want to accumulate warnings and errors from all the operations and return them.

I agree, and from what I've seen so far, I would prefer that to be the only use-case.

> I also looked at remaining two update calls in this PR, and thought they both were not good uses, so I removed them as well.

Ah, nice one. Given that chaining results is not actually implemented yet, it makes sense that there should be no existing usage of `Update` in these first commits. I looked at the initial implementation of these 2 new updates to make sure we're not missing anything and indeed, chaining is not relevant.

---

I've [posted](https://github.com/bitcoin/bitcoin/pull/25665#issuecomment-2079388456) my response to the second part of [your comment](https://github.com/bitcoin/bitcoin/pull/29906#issuecomment-2078452007) about `Update()` usage on #25665, since `Update()` is no longer in this PR.

furszy commented on src/validation.cpp:
```diff
@@ -970,11 +971,13 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
         }
-        ancestors = m_pool.CalculateMemPoolAncestors(*entry, cpfp_carve_out_limits);
-        if (!ancestors) return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "too-long-mempool-chain", error_message);
+        if (auto ancestors{m_pool.CalculateMemPoolAncestors(*entry, cpfp_carve_out_limits)}) {
```
This should be shadowing the previous `ancestors` variable (why the compiler isn't complaining about this?). The else path can also access the variable when it is declared inside the if statement.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1581128440

> But, in any case, it's not a big deal if you don't think it's worth it. I also don't think this is a problem anyway.

I'm happy to go with any suggestion. I don't see shadowing as a problem in all cases, so this is just the best way to write the code that I could think of.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1581580725

> I'm happy to go with any suggestion. I don't see shadowing as a problem in all cases, so this is just the best way to write the code that I could think of.

Not sure why I couldn't think of a good way to write this without shadowing last week. But added `_retry` suffixes now so the code should be clearer and shadowing should be gone :sunglasses:

achow101 commented:
ACK e1be443315be6ba6c80267e0e6be59deee77de50

achow101 commented on src/util/result.h:
```diff
@@ -39,13 +39,25 @@ class Result
+    //! Disallow copy constructor, require Result to be moved for efficiency.
+    Result(const Result&) = delete;
+
+    //! Disallow operator= to avoid confusion in the future when the Result
+    //! class gains support for richer error reporting, and callers should have
+    //! ability to set a new result value without clearing existing error
+    //! messages.
+    Result& operator=(const Result&) = default;
+    Result& operator=(Result&&) = default;
```
Can you briefly explain why these `operator=` are made private as opposed to deleting them?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1581409638

> Can you briefly explain why these `operator=` are made private as opposed to deleting them?

Changed these to `delete`. The default operators were needed in an earlier version of the PR which had an `Update()` method which called them privately.

ryanofsky commented on src/test/mempool_tests.cpp:
```diff
@@ -260,9 +262,12 @@ BOOST_AUTO_TEST_CASE(MempoolIndexingTest)
+        BOOST_REQUIRE(ancestors_calculated);
+        BOOST_CHECK(*ancestors_calculated == setAncestors);
+    }
+
```
re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1580997920

> nit: unnecessary extra newline

Thanks, removed

ryanofsky commented on src/wallet/test/fuzz/notifications.cpp:
```diff
@@ -106,13 +106,12 @@ struct FuzzedWallet {
+            return *Assert(wallet->GetNewChangeDestination(type));
         }
-        return *Assert(op_dest);
+        assert(false);
```
re: https://github.com/bitcoin/bitcoin/pull/29906#discussion_r1581039091

> nit: Is this line needed?

Nope, dropped this

ryanofsky reviewed (commented):
Updated e1be443315be6ba6c80267e0e6be59deee77de50 -> 9552619961049d7673f84c40917b0385be70b782 ([`pr/saferesult.7`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.7) -> [`pr/saferesult.8`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.7..pr/saferesult.8)) with small suggested cleanups

stickies-v commented:
Re-ACK 9552619961049d7673f84c40917b0385be70b782 , just addressing nits, no other changes

furszy reviewed (commented):
ACK 9552619961049d7673f84c40917b0385be70b782

sedited reviewed (approved):
ACK 9552619961049d7673f84c40917b0385be70b782

ryanofsky reviewed (commented):
Updated 9552619961049d7673f84c40917b0385be70b782 -> 6a8b2befeab25e4e92d8e947a23e78014695e06c ([`pr/saferesult.8`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.8) -> [`pr/saferesult.9`](https://github.com/ryanofsky/bitcoin/commits/pr/saferesult.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/saferesult.8..pr/saferesult.9)) tweaking variables to avoid shadowing

stickies-v commented:
re-ACK 6a8b2befeab25e4e92d8e947a23e78014695e06c

furszy reviewed (commented):
re-ACK https://github.com/bitcoin/bitcoin/commit/6a8b2befeab25e4e92d8e947a23e78014695e06c

achow101 commented:
ACK 6a8b2befeab25e4e92d8e947a23e78014695e06c
